### DISCUSSION TITLE
How to configure the expiration time of packages

### DISCUSSION THREAD

**Discussion by 89ao** [verdaccio/verdaccio#3488] [Q&A] [answered]
How long will the cached package exist for the longest time?
I am worried that there will be a situation where the upstream deletes the package but verdaccio do not delete it locally.
Situations that may cause illegal packages to be downloaded.
So please tell me how to configure the expiration/cache time of cached package,thank you.

**Comment by juanpicado** [comment, marked as answer]
Currently, there is no such thing, all is cached and you might want to remove them manually (yeah is not nice but just explaining how it is now), there are some initiatives for filtering packages and I think this might be useful to yeah I guess more help and point of views here might be need it to get it right and easy to apply. 

https://verdaccio.org/docs/next/dev-plugins/#filter-plugin

**Reply by 89ao** [reply]
on this situation : https://github.com/vuejs/vue-cli/issues/7054
Take node-ipc-9.2.2 for example ,once someone downloaded from our repository ,the package is cached locally.
Maby official-repo found the risk and deleted it , but node-ipc-9.2.2.tar.gz exist on our server, users still can download the threatened version my cause some problems.

Could you give some advices to solve this?

**Reply by juanpicado** [reply]
Well, luckily package managers are really good today:

- Pin your dependencies
- Use a lock file
- Use resolutions https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/ or https://yarnpkg.com/configuration/manifest#resolutions either pnpm or yarn any version (not sure about npm)
- Use audit tools, snyk, npm audit or others to avoid your lock file points to compromised versions

You can prevent what to fetch on the client side, that's the first line of defense, the registry can help but you should not rely on it, the public (npmjs) removed (MANUALLY) `colors` affected tarballs and versions a few months ago from the registry https://snyk.io/blog/open-source-npm-packages-colors-faker/  but https://registry.npmjs.org/colors still some hints remains in some fields

<img width="619" alt="Screenshot 2022-11-09 at 17 10 28" src="https://user-images.githubusercontent.com/558752/200881562-e76e272e-1a0a-4de3-87da-f7624fa6363e.png">

**Reply by 89ao** [reply]
Thanks to you @juanpicado  ,But lock specific package everytime by hand may not be a nicely way ?
Maybe verdaccio can automatic check packages exist or not on official/upstream site before someone download is a better way?
If official delete it ,automatelly delete it locally?

**Reply by juanpicado** [reply]
Well, here there is mo easy answer, you can hook to multiple registries, npmjs or other private ... how do you know was removed from A or B ? The registro A (nexus) might be still holding that data while npmjs does not. 

So npmjs is just the default upstream (uplink) but not the unique source of truth. 

Furthermore, a cache cannot be smart as you are describing, some packages might disappear from npmjs accidentally and suddenly removed from your side, yeah might be cached again but let's imagine package J was banned for X reason (not security) and you are not interested in being removed from your cache. 

I am sure there is more corner cases this manual removing is more accurate.

I thin the best way to go here is just avoiding those packages are served to the client either via client filter or maybe better verdaccio support for it in the server side. 

(Those are things I ve seen over years but I also looking for new points of views to make a better UX)

**Reply by 89ao** [reply]
Thank you all the same, these are indeed the problems. 
I'll try some ways to deleted them manually(still looking forward a MIRROR function of the incredible verdaccio~)
